core/src/main/scala/org/apache/spark/sql/delta/RowId.scala

```diff
@@ -87,4 +90,120 @@ object RowId {
     }
     latestMetadata
   }
+
+  /**
+   * Iterator to assign fresh row IDs to AddFiles in an iterator of actions.
+   */
+  private class RowIdAssigningIterator(
+      actions: Iterator[Action],
+      highWatermark: Option[Long]) extends Iterator[Action] {
+
+    val oldHighWatermark: Long = highWatermark.getOrElse(-1L)
+    var newHighWatermark: Long = oldHighWatermark
+
+    override def hasNext: Boolean = actions.hasNext || (newHighWatermark != oldHighWatermark)
```

Comment:
Maybe a small comment what the second part of the condition is for (emitting the new high watermark action)?

Reply:
Done